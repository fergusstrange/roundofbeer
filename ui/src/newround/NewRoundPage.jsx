import React from 'react';
import AddIcon from '@material-ui/icons/Add';
import { TextField } from '@material-ui/core';
import Fab from '@material-ui/core/Fab';
import PropTypes from 'prop-types';
import {
  actions, connect,
} from '../store/Store';

const NewRoundPage = ({ participant, participants }) => {
  const addParticipant = (e) => {
    e.preventDefault();
    if (participants.filter(p => p.toLowerCase().equals(participant.toLowerCase())).length === 0) {
      actions.addParticipant();
    }
  };

  const updateParticipant = (e) => {
    if (e.target.value) {
      actions.updateParticipant(e.target.value);
    }
  };

  return (
    <div>
      <div>
        {participants.map(name => (<div key={name}>{name}</div>))}
      </div>
      <div>
        <form onSubmit={addParticipant}>
          <div>
            <TextField label="name" value={participant} onChange={updateParticipant} />
            <Fab type="submit" color="primary" aria-label="Add" size="small">
              <AddIcon />
            </Fab>
          </div>
          <div>
            <Fab variant="extended" color="primary" aria-label="Add" onClick={actions.submitParticipants}>Start Round</Fab>
          </div>
        </form>
      </div>
    </div>
  );
};

NewRoundPage.propTypes = {
  participant: PropTypes.string.isRequired,
  participants: PropTypes.arrayOf(PropTypes.string).isRequired,
};


export default connect(state => state)(NewRoundPage);
