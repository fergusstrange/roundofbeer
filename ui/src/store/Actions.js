const actions = {
  updateError: error => async state => ({
    ...state,
    error,
  }),
  clearError: () => async state => ({
    ...state,
    error: undefined,
  }),
  updateRoundToken: data => async state => ({
    ...state,
    roundToken: data.token,
    round: undefined,
    participatingRounds: Array.isArray(state.participatingRounds)
      ? [...state.participatingRounds.filter(pr => pr.roundUrl !== data.roundUrl),
        {
          roundUrl: data.roundUrl,
          roundToken: data.token,
          participants: data.participants,
        }]
      : [{
        roundUrl: data.roundUrl,
        roundToken: data.token,
        participants: data.participants,
      }],
  }),
  updateRound: round => async state => ({
    ...state,
    round,
  }),
};

export default actions;
