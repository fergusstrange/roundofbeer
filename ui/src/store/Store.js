import { get, set } from 'local-storage';
import { createDakpan } from 'dakpan';
import ApiClient from '../client/Client';

const localStorageKey = 'roundOfBeerLocalStorage';
const client = new ApiClient();

function localStorageStateOrDefaults() {
  return get(localStorageKey)
    || {
      token: undefined,
      round: undefined,
      participant: '',
      participants: [],
    };
}

function updateLocalStore(state) {
  set(localStorageKey, state);
}

const [ContextProvider, useContext] = createDakpan(localStorageStateOrDefaults())({
  clearError: () => async state => ({
    ...state,
    error: undefined,
  }),
  addParticipant: () => ({ participant, ...state }) => ({
    ...state,
    participants: [
      ...state.participants,
      participant,
    ],
    participant: '',
  }),
  updateParticipant: participant => state => ({
    ...state,
    participant,
  }),
  submitParticipants: () => async (state) => {
    const { error, data } = await client.createRound(state.participants);
    return !error
      ? ({
        ...state,
        round: data.round,
        roundToken: data.token,
        participant: '',
        participants: [],
      })
      : ({
        ...state,
        error,
      });
  },
  fetchRound: () => async (state) => {
    const { error, data } = await client.fetchRound(state.roundToken);
    return error
      ? ({
        ...state,
        error,
      })
      : ({
        ...state,
        round: data.round,
        roundToken: data.token,
      });
  },
});

export { ContextProvider, useContext, updateLocalStore };
