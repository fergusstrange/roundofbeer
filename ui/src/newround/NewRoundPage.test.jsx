import React from 'react';
import renderer from 'react-test-renderer';
import NewRoundPage from './NewRoundPage';
import { ContextProvider } from '../store/Store';

describe('New Round Page Tests', () => {
  it('renders correctly', () => expect(renderer.create(
    <ContextProvider>
      <NewRoundPage history={{}} />
    </ContextProvider>,
  ).toJSON())
    .toMatchSnapshot());
});
