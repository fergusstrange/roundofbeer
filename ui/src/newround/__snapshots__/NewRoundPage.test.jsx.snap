// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Round Page Tests renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        className="MuiContainer-root MuiContainer-maxWidthLg"
      />
      <div
        className="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <div
          className="MuiFormControl-root MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
          >
            name
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={true}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value=""
            />
          </div>
        </div>
        <button
          aria-label="Add"
          className="MuiButtonBase-root MuiFab-root MuiFab-primary MuiFab-sizeSmall"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiFab-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <button
        aria-label="Add"
        className="MuiButtonBase-root MuiFab-root MuiFab-extended MuiFab-primary"
        disabled={false}
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiFab-label"
        >
          Start Round
        </span>
      </button>
    </div>
  </div>
</form>
`;
